test/controllers/sessions_controller_test.rb

```diff
@@ -0,0 +1,90 @@
+require "test_helper"
+
+class SessionsControllerTest < ActionDispatch::IntegrationTest
+  ##
+  # test all routes which lead to this controller
+  def test_routes
+    assert_routing(
+      { :path => "/login", :method => :get },
+      { :controller => "sessions", :action => "new" }
+    )
+    assert_routing(
+      { :path => "/login", :method => :post },
+      { :controller => "sessions", :action => "create" }
+    )
+    assert_recognizes(
+      { :controller => "sessions", :action => "new", :format => "html" },
+      { :path => "/login.html", :method => :get }
+    )
+
+    assert_routing(
+      { :path => "/logout", :method => :get },
+      { :controller => "sessions", :action => "destroy" }
+    )
+    assert_routing(
+      { :path => "/logout", :method => :post },
+      { :controller => "sessions", :action => "destroy" }
+    )
+    assert_recognizes(
+      { :controller => "sessions", :action => "destroy", :format => "html" },
+      { :path => "/logout.html", :method => :get }
+    )
+  end
+
+  def test_login
+    user = create(:user)
+
+    get login_path
+    assert_response :redirect
+    assert_redirected_to login_path(:cookie_test => true)
+    follow_redirect!
+    assert_response :success
+    assert_template "sessions/new"
```

Comment:
Presumably these template assertions could drop the `sessions` prefix now they been moved?

Reply:
This can go either way. The trouble with `assert_template "new"` is that it matches any template named `new` - it's not scoped to the controller under test. So if you end up rendering e.g. "users/new" somehow, the test will still pass. 

So two choices:
* Keep as-is, to ensure we are rendering the sign-in form and not e.g. the user registration form.
* Remove the scoping, to be consistent with all our other tests, and accept that the risk is minimal.

I'm happy either way.

Reply:
Let's leave it as it is in that case - we should probably aim to move to that syntax more generally!